### PR TITLE
Editorial for New Barns (usaco)

### PR DESCRIPTION
**Pull request by TheGamingMousse**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [x] I have linked this PR to any issues that it closes.

BTW, this problem is listed under three modules, but this solution is purely binlifting. Should I make that distinction or no?

### CODE REVIEW

**Comment by TheGamingMousse** [comment]
have not worked with images b4, hope it works lmfao

File: solutions/platinum/usaco-817.mdx [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+distance from a given node is to an endpoint of a diameter of our current tree.
+Now the problem boils down to keeping track of the diameter of each tree inside
+our graph after we add a node.
+
+Observe that if we add a leaf to the tree that becomes an endpoint of our
+new diameter, then the other endpoint of our new diameter is one of the endpoints
+of our old diameter. To understand why this is the case, consider laying out
+the tree in the following manner:
+
+![](<usaco-817/Diameter.png>)
```
Comment (SansPapyrus683) [inline solutions/platinum/usaco-817.mdx:34]: make the file name lowercase
maybe some alt text

File: solutions/platinum/usaco-817.mdx [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+	const int q, log2dist;
+	vector<vector<int>> up;
+	vector<pair<int, int>> diameter;
+	vector<int> root;
+	vector<int> dep;
+	int cur = -1;
+
+  public:
+	Tree(int _q)
+	    : q(_q), log2dist(1 + (int)log2(q)), up(log2dist, vector<int>(q)),
```
Comment (SansPapyrus683) [inline solutions/platinum/usaco-817.mdx:68]: just use `q`, no need for `_q`

File: solutions/platinum/usaco-817.mdx [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+				v = up[i][v];
+			}
+		}
+		return up[0][u];
+	}
+
+	int dist(int u, int v) { return dep[u] + dep[v] - 2 * dep[lca(u, v)]; }
+
+	void add(int par) {
+		++cur;
```
Comment (SansPapyrus683) [inline solutions/platinum/usaco-817.mdx:89]: postincrement

File: solutions/platinum/usaco-817.mdx [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+			int cur_dist = dist(a, b);
+			if (dist_a > cur_dist) {
+				diameter[root[cur]] = {a, cur};
+			} else if (dist_b > cur_dist) {
+				diameter[root[cur]] = {b, cur};
+			}
+		}
+	}
+
+	int qry(int u) {
```
Comment (SansPapyrus683) [inline solutions/platinum/usaco-817.mdx:113]: better name + doc comment

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
classes are truly a wonderful thing

**Comment by TheGamingMousse** [comment]
uhhhh now the image is broken crap

File: solutions/platinum/usaco-817.mdx [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+distance from a given node is to an endpoint of a diameter of our current tree.
+Now the problem boils down to keeping track of the diameter of each tree inside
+our graph after we add a node.
+
+Observe that if we add a leaf to the tree that becomes an endpoint of our
+new diameter, then the other endpoint of our new diameter is one of the endpoints
+of our old diameter. To understand why this is the case, consider laying out
+the tree in the following manner:
+
+![Example Graph](<usaco-817/diameter.png>)
```
Comment (ryanchou-dev) [inline solutions/platinum/usaco-817.mdx:34]: remove the brackets for file names

File: solutions/platinum/usaco-817.mdx [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+Say we add a child to node 11. We now have a new diameter starting from that
+child, and need to find a suitable endpoint. Note that it is always optimal
+to pick one endpoint of the current diameter because if there was a better
+option within one of the "subtrees", it would be the endpoint of the previous
+diameter.
+
+With this fact, we can easily maintain the diameter for each tree in our graph
+online. Finding the distance between each pair of nodes can be done using binary
+lifting.
+
```
Comment (ryanchou-dev) [inline solutions/platinum/usaco-817.mdx:45]: good explanation!

**Review by ryanchou-dev — COMMENTED** [review]
good! just fix the image rendering and we should be chill. lmk if u need any help with that 😁

File: solutions/platinum/usaco-817.mdx (hunk 1 above)
Comment (TheGamingMousse) [reply solutions/platinum/usaco-817.mdx]: done

File: solutions/platinum/usaco-817.mdx (hunk 6 above)
Comment (TheGamingMousse) [reply solutions/platinum/usaco-817.mdx:45]: :blobheart:

**Comment by TheGamingMousse** [comment]
> good! just fix the image rendering and we should be chill. lmk if u need any help with that 😁

ok image rendering now works when previewing so I think it works

File: solutions/platinum/usaco-817.mdx [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+using namespace std;
+
+class Tree {
+  private:
+	const int q, log2dist;
+	vector<vector<int>> up;
+	vector<pair<int, int>> diameter;
+	vector<int> root;
+	vector<int> dep;
+	int cur = -1;
```
Comment (SansPapyrus683) [inline solutions/platinum/usaco-817.mdx:64]: could you rename this to `curr` pwetty pwease :pleading_face:

Comment (TheGamingMousse) [reply solutions/platinum/usaco-817.mdx:64]: uh sure

File: solutions/platinum/usaco-817.mdx [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+	const int q, log2dist;
+	vector<vector<int>> up;
+	vector<pair<int, int>> diameter;
+	vector<int> root;
+	vector<int> dep;
+	int curr = -1;
+
+  public:
+	Tree(int q)
+	    : q(q), log2dist(1 + (int)log2(q)), up(log2dist, vector<int>(q)),
```
Comment (SansPapyrus683) [inline solutions/platinum/usaco-817.mdx:68]: ceil instead?

Comment (TheGamingMousse) [reply solutions/platinum/usaco-817.mdx:68]: eh i personally would rather use as little floating point as possible

Comment (SansPapyrus683) [reply solutions/platinum/usaco-817.mdx:68]: i mean unless you can come up with an example where ceil doesn't work
you already have bits/stdc++ and std
and it gets the idea across much better

**Review by ryanchou-dev — APPROVED** [review]

File: solutions/platinum/usaco-817.mdx (hunk 8 above)
Comment (TheGamingMousse) [reply solutions/platinum/usaco-817.mdx:68]: eh ok sure, i still dont like it but I can kind of see why

**Review by SansPapyrus683 — APPROVED** [review]
show proof of ac and we're fine

**Comment by TheGamingMousse** [comment]
image

*ahem* **merge the pull request now**

**Pull request merged**